csv: parse escape character in unquoted fields

fanyang01 opened the pull request:
Resolves #14440 

This PR enables DuckDB to parse the following TSV correctly:

```csv
0	\\	0
1	\		1
2	\
	2
3	a\	a	3
4	b\
b	4
5	c\\c	5
```

```sql
D FROM read_csv('t.txt', quote = '', escape = '\', sep = '\t', rfc_4180 = false) t (i, s, j);
┌───────┬─────────┬───────┐
│   i   │    s    │   j   │
│ int64 │ varchar │ int64 │
├───────┼─────────┼───────┤
│     0 │ \       │     0 │
│     1 │ \t      │     1 │
│     2 │ \n      │     2 │
│     3 │ a\ta    │     3 │
│     4 │ b\nb    │     4 │
│     5 │ c\c     │     5 │
└───────┴─────────┴───────┘
```

fanyang01 committed:
csv: parse escape character in unquoted fields

fanyang01 committed:
run format fix

pdet commented:
Hi @fanyang01  Thanks for the PR.

Could you add a couple more tests?
1. A test with escaped values within and without quoted values in the same file.
2. A test where the quoted and escaped values have the same character (e.g., quote = '"' and escape = '"')
3. A test where escaped unquoted values can happen in between CSV buffers (You might want a slightly bigger file for this test)
e.g., 
```sql
loop buffer_size 10 25

query ${number of columns}
read_csv('data/csv/unquoted_escape.csv', quote = '', escape = '\', sep = '\t')
----
${expected_result}

endloop
``` 
4. Also test with a file with at least 30k rows.

A bit orthogonal to this PR, but good to keep in mind is that it might also be possible to extend the sniffer to detect escaped in unquoted values.

fanyang01 commented:
@pdet I'm following your suggestion to add more tests. The reader fails for more complex cases and I'm working on fixing it now. Thanks for the feedback.

pdet commented:
> @pdet I'm following your suggestion to add more tests. The reader fails for more complex cases and I'm working on fixing it now. Thanks for the feedback.

No worries, many thanks for tackling this issue!

Also, one other test that should be added is with CSV Files with large (i.e., over 16 bytes) unquoted escaped values!

fanyang01 committed:
test: add csv files whose fields contain source code

fanyang01 commented:
> Also, one other test that should be added is with CSV Files with large (i.e., over 16 bytes) unquoted escaped values!

That is exactly what I'm experimenting with! I chose the [HumanEval](https://github.com/openai/human-eval) dataset, which contains 164 programming problems. Three fields of this dataset include multiline Python code, with multiple occurrences of `\n` and `,` characters. I also replaced the indentation with the tab character. I have addressed the failure, and DuckDB can now read the unquoted versions of them. Is this an acceptable test?

I will add the remaining tests soon.

fanyang01 committed:
Add more unquoted & escaped CSV tests

fanyang01 committed:
Run format fix

fanyang01 commented:
Hi @pdet,

I have added all the tests you suggested and successfully run them on my local machine.

fanyang01 commented:
The `read_csv(..., quote = '')` test failed for `'data/csv/quoted_values_delimited.csv'`:
```csv
"header1;header2;header3"
"value1;value2;value3"
"value1;value2;value3"
"value1;value2;value3"
```

It appears the issue is that the escape character is set to its default value of `'"'`, which is causing confusion for the parser. Changing the escape character to something else (e.g., `'\'`) could resolve this issue, but I am uncertain whether we should proceed with this change.

fanyang01 committed:
test: add escape character hint

fanyang01 committed:
test: add escape character hint for read_csv

fanyang01 committed:
test: add escape character hint for read_csv

fanyang01 committed:
csv: invalidate the state machine if escape char preceds ordinary char

fanyang01 committed:
Run format fix

fanyang01 commented:
Update:

The solution that minimizes the number of broken existing tests appears to be allowing the state machine to enter an invalid state when the escape character is followed by a non-special character. Otherwise, the sniffer may select undesired dialect candidates. For instance, it might choose `"` as the escape character when the quote is not `"`, leading to the entire file being parsed as a single line if each line ends with `"` (since the sniffer prefers a larger column count).

Additionally, for `\r\n` newlines, we must escape the both characters of the newline as a whole (otherwise the tests will fail on Windows). To handle this, I have introduced a new state, `ESCAPED_RETURN`, to represent the situation where the escape character is followed by a `\r`.

fanyang01 committed:
test: be newline-agnostic when comparing unqouted escaped csv

fanyang01 committed:
Merge branch 'duckdb:main' into csv-escape

fanyang01 committed:
test: be newline-agnostic when comparing unqouted escaped csv

fanyang01 committed:
test: escape `\n`

pdet commented on data/csv/unquoted_escape/human_eval.csv:
```diff
@@ -0,0 +1,5891 @@
+HumanEval/0,from typing import List\
+\
+\
+def has_close_elements(numbers: List[float]\, threshold: float) -> bool:\
```
I actually really like these files, since they seem like a pretty good stress test, but maybe we could keep them a bit smaller (say 10 lines) and try to use the buffer_size option to stress test their size, since they are also hard to read, debugging the full file might be painful in the future

fanyang01 replied:
Sure! I will truncate them to 10 rows.

fanyang01 replied:
Truncated to 10 rows.

pdet commented on data/csv/unquoted_escape/human_eval.tsv:
```diff
@@ -0,0 +1,5891 @@
+HumanEval/0	from typing import List\
+\
```
Same comment as the previous `human_eval` file

pdet commented on test/sql/copy/csv/unquoted_escape/basic.test:
```diff
@@ -0,0 +1,28 @@
+
+loop buffer_size 10 25
+
+query I
```
I would prefer to favor simple
`FROM read_csv()` queries when results are fairly small (say less than 100 rows), because they are much easier to debug if things go wrong in the future

pdet replied:
Same comment for the other test files

fanyang01 replied:
How can we make sqllogictest handle these special characters properly?

fanyang01 replied:
Certainly! I understand the importance of making the code future-proof and easy to maintain. I will make all the requested changes shortly.

fanyang01 replied:
- `basic.test`: Added a test case that uses simple `FROM read_csv` to read `plain.csv` (which does not contain tabs and newlines) and lets sqllogictest compare the results.
- `*.test`: Added comments to explain the CRLF -> LF replacement.

pdet commented on test/sql/copy/csv/test_sniff_csv.test:
```diff
@@ -94,7 +94,7 @@ FROM sniff_csv('data/csv/timings.csv')
 
 query IIIIIII
-FROM read_csv('data/csv/timings.csv', auto_detect=false, delim='|', quote='"', escape='\', new_line='\n', skip=0, header=true, columns={'tool': 'VARCHAR', 'sf': 'BIGINT', 'day': 'DATE', 'batch_type': 'VARCHAR', 'q': 'VARCHAR', 'parameters': 'VARCHAR', 'time': 'DOUBLE'}, dateformat='%Y-%m-%d') order by all limit 1;
+FROM read_csv('data/csv/timings.csv', auto_detect=false, delim='|', quote='"', escape='', new_line='\n', skip=0, header=true, columns={'tool': 'VARCHAR', 'sf': 'BIGINT', 'day': 'DATE', 'batch_type': 'VARCHAR', 'q': 'VARCHAR', 'parameters': 'VARCHAR', 'time': 'DOUBLE'}, dateformat='%Y-%m-%d') order by all limit 1;
```
This prompt is yielded by the `sniff_csv` function, from the test above it it seems that is returning `"\"` as a escape and not `""`

fanyang01 replied:
Reverted.

pdet reviewed (changes requested):
Many thanks for the great work here! I think the code looks excellent, and I just have some minor requests to make the tests a bit more straightforward.

My only concern now is that this change takes us a step away from the [RFC-4180](https://www.ietf.org/rfc/rfc4180.txt) specification.

I believe it’s most important to follow RFC-4180 when writing files, but in this case, the change introduces behavior that will alter how files are read structurally. Should we add a configuration option, perhaps? For example, `rfc_4180 = False` by default, where we aim to adhere to the specification as closely as possible.  What you think?

fanyang01 committed:
Recognize unquoted & escaped null string

fanyang01 commented:
Thank you for the detailed feedback! I’m glad to hear that you think the code looks good. Your elegant and well-structured CSV reader implementation makes it very easy to extend its functionality.

I appreciate your suggestions for the tests; I'll make those adjustments.

Maintaining compliance to RFC and backward compatibility is indeed important. Adding a configuration option like rfc_4180 = True/False sounds like a great idea.  I’ll implement this change.

Additionally, I have some [pending changes](https://github.com/duckdb/duckdb/compare/main...fanyang01:duckdb:escape-null-str) regarding unquoted and escaped null strings (such as MySQL's `\N`). Should we include those in this PR?

fanyang01 committed:
feat: add RFC 4180 conformance option for CSV reading

fanyang01 committed:
Run format fix

fanyang01 committed:
Test handling for unquoted escaped null strings in CSV

fanyang01 committed:
Merge branch 'main' into csv-escape

fanyang01 committed:
Run format fix

fanyang01 committed:
Update nodes.json

fanyang01 commented:
I have added the global `rfc_4180` option and set it to default to `true` to maintain backward compatibility. When `rfc_4180` is set to `false`, the reader will recognize and remove the escape character in unquoted fields. Additionally, to recognize MySQL-style null strings (`\N`), the state machine has been relaxed (compared to the previous implementation) to accept any character following an escape. Please let me know if you have any concerns.

fanyang01 commented:
I've just learned that PostgreSQL's `COPY TO ... FORMAT TEXT` behaves similarly to MySQL's equivalent. This PR will therefore allow us to emulate the behavior of PostgreSQL's `COPY FROM ... FORMAT TEXT`, with the exception of recognizing octal or hex-digit backslash sequences—which, in any case, aren't emitted by `COPY TO`.

pdet commented on src/execution/operator/csv_scanner/scanner/string_value_scanner.cpp:
```diff
@@ -239,31 +239,39 @@ void StringValueResult::AddValueToVector(const char *value_ptr, const idx_t size
+		const bool check_unquoted_escaped_null =
+		    state_machine.state_machine_options.rfc_4180.GetValue() == false && escaped && !quoted && size == 1;
+		for (idx_t i = 0; i < null_str_count; i++) {
+			bool isNull = false;
```
nit: is_null

fanyang01 replied:
Done. I forgot it's C++ :sweat_smile:

pdet commented on src/execution/operator/csv_scanner/state_machine/csv_state_machine_cache.cpp:
```diff
@@ -202,6 +251,7 @@ void CSVStateMachineCache::Insert(const CSVStateMachineOptions &state_machine_op
 	transition_array.skip_standard[static_cast<uint8_t>('\n')] = false;
 	transition_array.skip_standard[static_cast<uint8_t>('\r')] = false;
 	transition_array.skip_standard[comment] = false;
+	transition_array.skip_standard[escape] = false;
```
Should this be wrapped on the `rfc_4180` flag?

fanyang01 replied:
Done

pdet reviewed (changes requested):
Thanks so much for the hard work here, it looks excellent,

Just two minor extra comments, and I think it's good to go!

fanyang01 committed:
Do not skip escape character in standard state only if rfc_4180 = false

fanyang01 committed:
isNull -> is_null

fanyang01 committed:
Merge branch 'feature' into csv-escape

fanyang01 committed:
Run format fix

fanyang01 committed:
Fix EnumUtil

fanyang01 commented:
I apologize for previously selecting the main branch as the base. I have now retargeted this PR to the feature branch.

fanyang01 committed:
Merge branch 'main' into csv-escape

fanyang01 commented:
Oops, it appears I should target the main branch because the new option needs to come after the `encoding` option introduced in https://github.com/duckdb/duckdb/pull/14560.

pdet commented:
Hi @fanyang01 I think it all looks excellent! Indeed we merged `feature` with `main` yesterday, so it should be now targeting`main`!

Thanks again for all the effort you put in this PR and for going over my comments and nits :-)

It looks good to me. 

@Mytherin maybe you would like to have a look?

pkoppstein commented:
@fanyang01, @pdet, @Mytherin - I hope it's not too late to comment 
on the choice of name for the new flag, namely `rfc_4180`.

The choice of name seems to be at confusing, because 
RFC4180 bundles together several requirements, and does
so in a rather unsatisfactory way. Using this particular name 
drags in a whole bunch of dirty laundry that can only add to
confusion and/or complexity.

pkoppstein commented:
@fanyang01 - Unfortunately, at this point I don't understand the new rfc_4180 flag
well enough to be able to offer a helpful alternative name.  In fact,
your description ("remove the escape character in unquoted fields")
leaves me wondering whether the effect you're trying to achieve might
not be achieved in some other way with fewer complications.

I think part of the problem here is that in RFC 4180 and similar
interpretations of CSV, the doubling of DQUOTE was not intended
to define a general "escape character" mechanism.  The relevant line from the RFC is:

   escaped = DQUOTE *(TEXTDATA / COMMA / CR / LF / 2DQUOTE) DQUOTE

That is, as far as RFC 4180 is concerned, there are two kinds of
fields: "escaped fields" and "ordinary fields".  There is no mention
of an "escape character", so as soon as we introduce that as a
separate idea, complications start to pile up.

I suspect that DuckDB's approach to supporting multiple variations of
multiple CSV dialects would in the end have been better served by having
separate flags for an "escape character" (in the usual sense of a
single character like '\'), and the *policy* of doubling the "quote"
character (whatever it might be) in an "escaped field".

---

As for "dialects" of CSV - yes, I think that a flag that
would clearly specify a certain behavior and how it is affected
by all the feature-specific flags would be an excellent way to proceed:

* using a flag name such as "dialect" (or "mode" or "defaults")
would make it clear that only one dialect at a time can be specified;
* the documentation would clarify exactly which flag values it implies;
* the semantics would also be clear enough (i.e. explicit per-feature
flags would always over-ride the "dialect" setting).

The question in my mind is whether or not dialect=rfc_4180 would 
be equivalent to a constellation of feature-specific flags.
At first I was inclined to think that such "dialect" flags
should be no more than such a constellation, but I'm worried
that there are too many complications. As an exercise, though, it would
probably pay off to see what additional flags would be needed, and
how they'd interact with each other.

An alternative to adding `dialects` that you've probably already considered would be
to offer more than just read_csv() and read_csv_auto() functions.
I've often wished it were easier to read and write TSV files in accordance with
a particular set of requirements.

Anyway, to summarize, my hope is that DuckDB will support
a small handful of dialects by providing a clear exposition of exactly what
each one means, especially in relation to the other feature-specific flags.
